src/cpu/intel.rs

```diff
@@ -98,47 +94,127 @@ pub(super) mod featureflags {
-        | if cfg!(target_feature = "sse4.1") { super::Sse41::mask() } else { 0 }
-        | if cfg!(target_feature = "ssse3") { super::Ssse3::mask() } else { 0 }
+        | if cfg!(target_feature = "sse4.1") { Sse41::mask() } else { 0 }
+        | if cfg!(target_feature = "ssse3") { Ssse3::mask() } else { 0 }
         ;
 
     pub const FORCE_DYNAMIC_DETECTION: u32 = 0;
 }
 
-fn cpuid_to_caps_and_set_c_flags(cpuid: &[u32; 4]) -> u32 {
-    // "Intel" citations are for "Intel 64 and IA-32 Architectures Software
-    // Developer’s Manual", Combined Volumes, December 2024.
-    // "AMD" citations are for "AMD64 Technology AMD64 Architecture
-    // Programmer’s Manual, Volumes 1-5" Revision 4.08 April 2024.
+struct CpuidSummary {
+    #[cfg(target_arch = "x86_64")]
+    is_intel: bool,
+    leaf1_edx: u32,
+    leaf1_ecx: u32,
+    #[cfg(target_arch = "x86_64")]
+    extended_features_ecx: u32,
+    #[cfg(target_arch = "x86_64")]
+    extended_features_ebx: u32,
+    xcr0: u64,
+}
+
+// SAFETY: This unconditionally uses CPUID because we don't have a good
+// way to detect CPUID and because we don't know of a CPU that supports
+// SSE2 (that we currently statically require) but doesn't support
+// CPUID. SGX is one environment where CPUID isn't allowed but where
+// SSE2 is statically supported. Ideally there would be a
+// `cfg!(target_feature = "cpuid")` we could use.
+unsafe fn cpuid_all() -> CpuidSummary {
+    #[cfg(target_arch = "x86")]
+    use core::arch::x86 as arch;
+    #[cfg(target_arch = "x86_64")]
+    use core::arch::x86_64 as arch;
 
-    // The `prefixed_extern!` uses below assume this
+    // MSRV(1.66) avoids miscompilations when calling `__cpuid`;
+    // see https://github.com/rust-lang/rust/pull/101861.
+
+    // Intel: "21.1.1 Notes on Where to Start".
+    let r = unsafe { arch::__cpuid(0) };
+
+    let leaf1_edx;
+    let leaf1_ecx;
+
+    #[cfg(target_arch = "x86_64")]
+    let is_intel = (r.ebx == 0x756e6547) && (r.edx == 0x49656e69) && (r.ecx == 0x6c65746e);
+
+    #[cfg(target_arch = "x86_64")]
+    let (extended_features_ecx, extended_features_ebx);
+
+    if r.eax >= 1 {
+        // SAFETY: `r.eax >= 1` indicates leaf 1 is available.
+        let r = unsafe { arch::__cpuid(1) };
+        leaf1_edx = r.edx;
+        leaf1_ecx = r.ecx;
+
+        #[cfg(target_arch = "x86_64")]
+        if r.eax >= 7 {
+            // SAFETY: `r.eax >= 7` implies we can execute this.
+            let r = unsafe { arch::__cpuid(7) };
+            extended_features_ecx = r.ecx;
+            extended_features_ebx = r.ebx;
+        } else {
+            extended_features_ecx = 0;
+            extended_features_ebx = 0;
+        }
+    } else {
+        // Expected to be unreachable on any environment we currently
+        // support.
+        leaf1_edx = 0;
+        leaf1_ecx = 0;
+        #[cfg(target_arch = "x86_64")]
+        {
+            extended_features_ecx = 0;
+            extended_features_ebx = 0;
+        }
+    }
```

Comment:
As codecov points out, and the comment notes, these lines are unreachable in *unpatched* versions of *ring*. (Some people do patch it in ways that make these lines reachable.)